shared/src/lib.rs

```diff
@@ -75,22 +91,101 @@ impl From<SocketAddr> for PravegaNodeUri {
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => {
+                let mut addrs_vec: Vec<_> =
+                    format!("{}:{}", uri_parts.domain_name.unwrap(), uri_parts.port.unwrap())
+                        .to_socket_addrs()
+                        .expect("Unable to resolve domain")
+                        .collect();
+                addrs_vec.pop().expect("get the first SocketAddr")
+            }
+            Err(e) => panic!("{}", e),
+        }
     }
 
     pub fn domain_name(&self) -> String {
-        let parts: Vec<_> = self.0.split(':').collect();
-        parts[0].to_string()
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => uri_parts.domain_name.expect("uri missing domain name"),
+            Err(e) => panic!("{}", e),
+        }
     }
 
     pub fn port(&self) -> u16 {
-        let parts: Vec<_> = self.0.split(':').collect();
-        parts[1].parse::<u16>().expect("parse port to u16")
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(uri_parts) => uri_parts.port.expect("parse port to u16"),
+            Err(e) => panic!("{}", e),
+        }
+    }
+
+    /// Return Result of the uri scheme or empty string if no scheme was specified
+    pub fn scheme(&self) -> PravegaNodeUriParseResult<String> {
+        match PravegaNodeUri::uri_parts_from_string(self.to_string()) {
+            Ok(sa) => match sa.scheme {
+                Some(scheme) => Ok(scheme),
+                _ => Ok("".to_string()),
+            },
+            Err(e) => Err(e),
+        }
+    }
+
+    /// verifies the uri is well-formed (contains at least host:port, with optional scheme:// prefix)
+    ///
+    pub fn is_well_formed(uri: String) -> bool {
+        match PravegaNodeUri::uri_parts_from_string(uri) {
+            Ok(uri_parts) => uri_parts.port.is_some() && uri_parts.domain_name.is_some(),
+            Err(_) => false,
+        }
+    }
+
+    fn uri_parts_from_string(uri: String) -> PravegaNodeUriParseResult<PravegaNodeUriParts> {
+        let mut uri_parts: PravegaNodeUriParts = PravegaNodeUriParts::default();
+
+        // The Java client supports multiple comma separated endpoints in a single string
+        // where the first endpoint has the scheme to be applied to all endpoints.
+        // To be semi-compatible with the Java code this method accepts a string containing comma separated
+        // endpoints, but it uses only the first endpoint in the string
+        let first_endpoint = match uri.split(',').next() {
+            Some(endpoint) => endpoint,
+            _ => {
+                return Err(PravegaNodeUriParseError::ParseError {
+                    error_msg: format!("malformed uri {}", uri),
+                })
+            }
+        };
+
+        let mut scheme_authority: Vec<_> = first_endpoint.split("://").collect();
```

Comment:
While by no means a requirement, I'd be tempted to use [regex](https://docs.rs/regex/1.5.4/regex/):
```regex
(?x)
((?:P<scheme>[[:alnum:]-]*)://)?
(?:P<host>[[:alnum:]-]*)):
(?:P<port>[[:digit:]]*)
```

Reply:
Hi Tom,

In my original PR comment I said:

> I considered using Regex but its currently only used in the auth crate, so didn't seem worth pulling into this one.

At that time I wasn't sure what the ramifications were of using regex in this crate.

I will go ahead and change it.